### PR TITLE
Add qsimcirq recipe

### PR DESCRIPTION
**Pull request by basnijholt**
Closes https://github.com/conda-forge/staged-recipes/issues/22374

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

Resources (examples) I used:
- https://github.com/conda-forge/cuda-version-feedstock/blob/main/recipe/README.md
- https://github.com/conda-forge/rmm-feedstock/blob/main/recipe/meta.yaml
- https://github.com/conda-forge/cuda-python-feedstock/blob/main/recipe/meta.yaml
- https://github.com/conda-forge/librealsense-feedstock/pull/39/files
- https://github.com/conda-forge/librealsense-feedstock/blob/main/recipe/build.sh
- https://github.com/conda-forge/cuquantum-feedstock

### CODE REVIEW

**Comment by basnijholt** [comment]
@leofang and @jakirkham, you both seem to be the experts when it comes to compiling conda packages with CUDA support.

After lots of trial and error I was able to compile [the `qsim` package](https://github.com/quantumlib/qsim) with CUDA. Would you be able to take a look at what I have done?

Thanks in advance! Feel free to ignore and my apologies for tagging you without warrant.

**Comment by leofang** [comment]
Thanks, I will try to get back tomorrow. I think there are some changes needed.

**Comment by basnijholt** [comment]
@conda-forge/help-c-cpp and @conda-forge/help-python, this one is ready for review, thanks 🙏😄

**Comment by basnijholt** [comment]
I have removed CUDA for now to make the review process faster.

All CI pipelines are passing ✅ and hopefully someone (@conda-forge/help-c-cpp and @conda-forge/help-python) can chime in.

I am kind of blocked by this ATM, so if anyone can find a few minutes, it would be greatly appreciated  🙏

File: recipes/qsimcirq/meta.yaml
Comment (basnijholt) [inline recipes/qsimcirq/meta.yaml:64]: @BastianZim, could you verify here again that you want to be a maintainer (like done here https://github.com/conda-forge/cirq-feedstock/pull/13#issuecomment-1828689789)?

Comment (BastianZim) [reply recipes/qsimcirq/meta.yaml:64]: Yes, I'm happy to!

Comment (basnijholt) [reply recipes/qsimcirq/meta.yaml:64]: Thanks @BastianZim!

All builds are passing. Let's get this in ASAP 😄

File: recipes/qsimcirq/meta.yaml
Comment (BastianZim) [inline recipes/qsimcirq/meta.yaml:9]:
```suggestion
  url: https://github.com/quantumlib/qsim/archive/refs/tags/v{{ version }}.tar.gz
```

File: recipes/qsimcirq/meta.yaml
Comment (BastianZim) [inline recipes/qsimcirq/meta.yaml:13]:
```suggestion
```
We don't build for python2 anymore.

File: recipes/qsimcirq/meta.yaml
Comment (BastianZim) [inline recipes/qsimcirq/meta.yaml:40]: I don't see a numpy in host or build. This needs to be in both.

**Review by BastianZim — CHANGES_REQUESTED** [review]
I'm not completely familiar with python-c but noticed a couple of small things.

File: recipes/qsimcirq/meta.yaml
Comment (basnijholt) [reply recipes/qsimcirq/meta.yaml:40]: Thanks! Added in https://github.com/conda-forge/staged-recipes/pull/24504/commits/10dedc4a4c03c4c99cba8adbf0a1e0662d37bd33.

File: recipes/qsimcirq/meta.yaml
Comment (basnijholt) [reply recipes/qsimcirq/meta.yaml:36]: I do not think it is needed but they list it as requirement: https://github.com/quantumlib/qsim/blob/master/requirements.txt

Comment (basnijholt) [reply recipes/qsimcirq/meta.yaml]: Thanks a lot for taking a look here!

I just tried to build it without `pybind11` but this means that `pip check` [will fail](https://github.com/conda-forge/staged-recipes/runs/19350310895).
So I added it back.

File: recipes/qsimcirq/meta.yaml
Comment (basnijholt) [reply recipes/qsimcirq/meta.yaml]: Just moved this to build. Thanks!

All builds are green again ✅

**Review by BastianZim — APPROVED** [review]

**Comment by basnijholt** [comment]
@conda-forge/help-c-cpp and @conda-forge/help-python, could someone merge this?

@BastianZim (who has merging rights) has approved this one, but since I have added him as a `recipe-maintainer`, policy dictates he cannot merge.

File: recipes/qsimcirq/meta.yaml
Comment (xhochy) [reply recipes/qsimcirq/meta.yaml]: Then report this as an issue upstream. It should only be listed as an setup requirement.

**Review by xhochy — APPROVED** [review]

**Pull request merged**
